### ISSUE TITLE
Implement ERC734

### ISSUE THREAD

**Issue by tyleryasaka** [ERC725Alliance/ERC725#5]
It would be great to have an implementation of ERC734 in this repo (just like we have an ERC725 implementation [here](https://github.com/ERC725Alliance/erc725/tree/master/contracts/contracts)). The current draft of the ERC734 spec is [here](https://github.com/ERC725Alliance/erc725/blob/master/docs/ERC-734.md). Please comment if you're interested in working on this and have any questions.

**Comment by filiplazovic** [comment]
I would like to work on this

**Comment by CPerezz** [comment]
I would like to join it too! Should we talk a bit before starting? @filiplazovic  and @tyleryasaka ?

**Comment by tyleryasaka** [comment]
Awesome! @CPerezz it might be hard to find a time that works for all of us to discuss over a call, but we can definitely discuss in this thread.

So to make sure we're on the same page, the main idea is that the key manager (734) should be associated with an ERC725 identity contract, and will pass along the *execute* parameters to this contract. I did something very similar [here](https://github.com/tyleryasaka/identity-proposals/blob/master/contracts/KeyManager.sol#L110).

Other than that, the rest should look pretty much like the former ERC725 specs. Here are two implementations of this:
- https://github.com/mirceapasoi/erc725-735
- https://github.com/OriginProtocol/origin-playground

The [ERC734 doc](https://github.com/ERC725Alliance/erc725/blob/master/docs/ERC-734.md) provides the exact specs, although it is of course a work in progress. It hasn't been proofread either, so there may be mistakes and typos. Please feel free to point out any errors or areas of confusion in that doc and I will try to fix them.

Whoever begins working on this: my suggestion would be to implement only a small part of the spec (maybe 2 or 3 functions) and then submit that for review so that we can all discuss and make sure we're on the same page.

**Comment by filiplazovic** [comment]
For now i have few things:

- Can a key have multiple purposes? In specs `struct Key` has `uint256[]`, but function `addKey` accepts `uint256`

- Why are we passing purpose to `removeKey`?

**Comment by tyleryasaka** [comment]
@filiplazovic Good question. Yes, I think keys can have multiple purposes. To assign multiple purposes to a key, you would have to call `addKey` multiple times, once for each purpose. (So if you want to add a purpose to an existing key, you would call `addKey` with the existing key and the new purpose.)

Likewise, `removeKey` does not necessarily remove the entire key. It just removes the specified key/purpose combination. So if you have a key with 3 purposes, you would have to call `removeKey` 3 times to completely remove it.

Does that make sense? I'm not sure if this is the best way to do it, but that is how I interpret the current standard.

**Comment by filiplazovic** [comment]
After looking at this in detail, i thinks having a purpose as an array creates some problems here. I understand that it allows us to return all purposes easily in `getKey`, but items of the array can be added and deleted multiple times, and maintaining length and order of items can be tricky and expensive in solidity.

If we turn purposes to `mapping`, it would make it a lot of easier to add and remove them. This would require new functions `addKeyPurpose` and `removeKeyPurpose` which would separate logic for creating a key and adding/removing purposes for that keys. A drawback here is that we can't return all purposes in `getKey` (we can't return the whole mapping), instead the client needs to call `keyHasPurpose` for every purpose separately.

Also, i think, the filtering that `getKeysByPurpose` would provide, should be done on the client-side as it would probably require another mapping to pack the data in this particular way.

@tyleryasaka Let me know what you think

**Comment by tyleryasaka** [comment]
@filiplazovic I tend to agree with you. I previously suggested to @frozeman that we drop the `getKeysByPurpose` method. I will try to get his feedback here.

In the meantime, feel free to go ahead and implement the contracts the way you described. I think once you have working code it will be easier to update the standards as necessary.

**Comment by filiplazovic** [comment]
Alternatively, we can keep the array, but add function `setKeyPurposes` which will accept array to set all purposes at once. So instead of adding purposes one by one, we can set the whole array with one transaction. The downside would be that if we want to remove just one purpose out of 5, we have to pass new array of 4 to replace it.

With this approach`keyHasPurpose` could not work without a loop.
Then if we want to check if key has specific purpose inside some function, we need to pass index of the purpose we are checking, e.g. for `approve` function:
```
function approve(uint256 _id, bool _approve, uint256 _purposeIndex) {
   require(
       keys[keccak256(msg.sender)].purposes[_purposeIndex] == 2,
       "Sender does not have action key"
    );
}
```

So to sum up the characteristics for both solutions:

First solution:
   - use mapping for purposes
   - read purposes by calling `keyHasPurpose` for every purpose
   - easy checking if some key has particular purpose
   - Adding and removing purposes one by one (or multiple additions and multiple removals, but adding and removing needs to be separate)

Second solution:
  - use array
  - read all purposes by calling `getKey`
  - caller needs to pass additional parametar so we can check if key has particular purpose (to avoid 
  looping)
  - all purposes can be changed in one transaction

**Comment by tyleryasaka** [comment]
@filiplazovic I personally prefer the first solution (using a mapping). It's simpler.

**Comment by jordipainan** [comment]
Hello! I have been reading along the thread and I think the same. It is simpler and easier to handle a mapping than an array. Also I am agree with having multiple purposes in keys and having only the ```keyHasPurpose``` method. ```getKeysByPurpose``` seems to be a feature that comes to fulfill the same purpose that ```keyHasPurpose``` :). One of the two methods can satisfy almost the same, at least in my opinion.

**Comment by CPerezz** [comment]
@filiplazovic Perfect! I've been following the thread and meanwhile implementing the payments through the  Proxy with @jordipainan . You've done a great work mate!

**Comment by frozeman** [comment]
Thanks @tyleryasaka got the implementation.

TODO:
- [ ] The implementation should be more extensible tested.
- [ ] The implementation can make use of solidity 0.6.1 feature and return arrays instead of bit masks
- [ ] The implementation is missing functions that need implementation

**Comment by YamenMerhi** [comment]
ERC734 is now deprecated, you can use [LSP6-KeyManager](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-6-KeyManager.md) for a Key Manager system to control ERC725 contracts.
